File: core/src/com/bombbird/terminalcontrol/entities/procedures/HoldProcedure.java
package com.bombbird.terminalcontrol.entities.procedures;

import com.badlogic.gdx.utils.Array;
import com.bombbird.terminalcontrol.entities.sidstar.Star;
import com.bombbird.terminalcontrol.entities.waypoints.Waypoint;

import java.util.HashMap;

public class HoldProcedure {
    private HashMap<String, HoldingPoints> holdingPoints;

    public HoldProcedure() {
        holdingPoints = new HashMap<String, HoldingPoints>();
    }

    public HoldProcedure(Star star) {
        this();

        Array<Waypoint> waypoints = star.getWaypoints();
        for (int i = 0; i < waypoints.size; i++) {
            String wptName = waypoints.get(i).getName();
            HashMap<String, HoldingPoints> holdingPoint = star.getAirport().getHoldingPoints();
            if (holdingPoint.containsKey(wptName)) {
                holdingPoints.put(wptName, holdingPoint.get(wptName));
            }
        }
    }

    public int getEntryProcAtWpt(Waypoint waypoint, double heading) {
        return holdingPoints.get(waypoint.getName()).getEntryProc(heading);
    }

    public void renderShape(Waypoint waypoint) {
        holdingPoints.get(waypoint.getName()).renderShape();
    }

    public int getMaxSpdAtWpt(Waypoint waypoint) {
        return holdingPoints.get(waypoint.getName()).getMaxSpd();
    }

    public int getInboundHdgAtWpt(Waypoint waypoint) {
        return holdingPoints.get(waypoint.getName()).getInboundHdg();
    }

    public float getLegDistAtWpt(Waypoint waypoint) {
        return holdingPoints.get(waypoint.getName()).getLegDist();
    }

    public float[] getOppPtAtWpt(Waypoint waypoint) {
        return holdingPoints.get(waypoint.getName()).getOppPoint();
    }

    public boolean isLeftAtWpt(Waypoint waypoint) {
        return holdingPoints.get(waypoint.getName()).isLeft();
    }

    public int[] getAltRestAtWpt(Waypoint waypoint) {
        return holdingPoints.get(waypoint.getName()).getAltRestrictions();
    }

    public HashMap<String, HoldingPoints> getHoldingPoints() {
        return holdingPoints;
    }
}

File: core/src/com/bombbird/terminalcontrol/entities/procedures/holding/HoldProcedure.kt
package com.bombbird.terminalcontrol.entities.procedures.holding

import com.bombbird.terminalcontrol.entities.sidstar.Star
import com.bombbird.terminalcontrol.entities.waypoints.Waypoint
import java.util.*

class HoldProcedure() {
    val holdingPoints: HashMap<String, HoldingPoints> = HashMap()

    constructor(star: Star) : this() {
        val waypoints = star.waypoints
        for (i in 0 until waypoints.size) {
            val wptName = waypoints[i].name
            val holdingPoint = star.airport.holdingPoints
            holdingPoint[wptName]?.let { holdingPoints[wptName] = it }
        }
    }

    fun getEntryProcAtWpt(waypoint: Waypoint, heading: Double): Int {
        return holdingPoints[waypoint.name]?.getEntryProc(heading) ?: 3
    }

    fun renderShape(waypoint: Waypoint) {
        holdingPoints[waypoint.name]?.renderShape()
    }

    fun getMaxSpdAtWpt(waypoint: Waypoint): Int {
        return holdingPoints[waypoint.name]?.maxSpd ?: 250
    }

    fun getInboundHdgAtWpt(waypoint: Waypoint): Int {
        return holdingPoints[waypoint.name]?.inboundHdg ?: 360
    }

    fun getLegDistAtWpt(waypoint: Waypoint): Float {
        return holdingPoints[waypoint.name]?.legDist ?: 5f
    }

    fun getOppPtAtWpt(waypoint: Waypoint): FloatArray {
        return holdingPoints[waypoint.name]?.oppPoint ?: floatArrayOf(0f, 0f)
    }

    fun isLeftAtWpt(waypoint: Waypoint): Boolean {
        return holdingPoints[waypoint.name]?.isLeft == true
    }

    fun getAltRestAtWpt(waypoint: Waypoint): IntArray {
        return holdingPoints[waypoint.name]?.altRestrictions ?: intArrayOf(-1, -1)
    }
}